cxx/emissions/bigram_string.cc

```diff
@@ -0,0 +1,337 @@
+  return old_cost - log_prob;
+}
+
+void BigramStringEmission::incorporate(
+    const std::pair<std::string, std::string>& x, double weight) {
+  N += weight;
+
+  std::vector<StrAlignment> alignments;
+
+  topk_alignments(10, x.first, x.second,
+                  [&](const StrAlignment &a, double old_cost) {
+                    return log_prob_distance(a, old_cost);
+                  },
+                  &alignments);
+
+  double total_prob = 0.0;
+  for (auto& a : alignments) {
+    a.cost = exp(a.cost);  // Turn all costs into non-log probabilities
+    total_prob += a.cost;
+  }
+
+  for (const auto& a : alignments) {
+    double w = weight * a.cost / total_prob;
+    std::string insertion_context = "";
+    for (auto it = a.align_pieces.begin();
+         it != a.align_pieces.end();
+         ++it) {
+      size_t icontext = get_index(insertion_context);
+      switch (it->index()) {
+        case 0:  // Deletion
+          // which is actually no insertion, then a deletion.
+          {
+            char deleted_char = std::get<Deletion>(*it).deleted_char;
+            insertions[icontext].incorporate(0, w);
+            substitutions[get_index(deleted_char)].incorporate(0, w);
+            insertion_context = std::string() + deleted_char;
+          }
+          break;
+        case 1:  // Insertion
+          {
+            char inserted_char = std::get<Insertion>(*it).inserted_char;
+            insertions[icontext].incorporate(get_index(inserted_char), w);
+          }
+          break;
+        case 2:  // Substitution
+          // which is actually no insertion, then a substitution.
+          {
+            char original_char = std::get<Substitution>(*it).original;
+            char replacement_char = std::get<Substitution>(*it).replacement;
+            insertions[icontext].incorporate(0, w);
+            substitutions[get_index(original_char)].incorporate(get_index(replacement_char), w);
+            insertion_context = std::string() + original_char;
+          }
+          break;
+        case 3:  // Match
+          // which is actually no insertion, then a match.
+          {
+            char c = std::get<Match>(*it).c;
+            insertions[icontext].incorporate(0, w);
+            substitutions[get_index(c)].incorporate(get_index(c), w);
+            insertion_context = std::string() + c;
+          }
+          break;
+      }
+    }
+  }
+}
+
+double BigramStringEmission::logp(
+    const std::pair<std::string, std::string>& x) const {
+  // Store state.
+  double old_N = N;
+  std::vector<DirichletCategorical> orig_insertions;
+  std::vector<DirichletCategorical> orig_substitutions;
+  std::copy(insertions.begin(), insertions.end(), std::back_inserter(orig_insertions));
+  std::copy(substitutions.begin(), substitutions.end(),
+            std::back_inserter(orig_substitutions));
+  double orig_lp = logp_score();
+  BigramStringEmission* unsafe_this = const_cast<BigramStringEmission*>(this);
```

Comment:
Can we avoid this, maybe by enabling the copy constructor and modifying a copy instead? Or by declaring insertions/substitutions `mutable`, similar to the bigram distribution?

Reply:
Excellent idea!  I enabled the copy constructor and now it is so much nicer!!